transport/src/main/java/io/netty/channel/DefaultChannelPipeline.java

```diff
@@ -462,46 +611,80 @@ public ChannelHandler replace(String oldName, String newName, ChannelHandler new
 
     private ChannelHandler replace(
             final AbstractChannelHandlerContext ctx, String newName, ChannelHandler newHandler) {
-
         assert ctx != head && ctx != tail;
 
-        Future<?> future;
+        final AbstractChannelHandlerContext newCtx;
+        final EventExecutor executor;
+        final boolean inEventLoop;
         synchronized (this) {
+            checkMultiplicity(newHandler);
+
             if (newName == null) {
                 newName = ctx.name();
             } else if (!ctx.name().equals(newName)) {
                 newName = filterName(newName, newHandler);
             }
 
-            final AbstractChannelHandlerContext newCtx =
-                    new DefaultChannelHandlerContext(this, ctx.invoker, newName, newHandler);
-
-            if (!isExecuteLater(newCtx)) {
+            newCtx = new DefaultChannelHandlerContext(this, ctx.invoker, newName, newHandler);
+            executor = executorSafe(ctx.invoker);
+            if (executor == null) {
                 replace0(ctx, newCtx);
+                callHandlerCallbackLater(newCtx, true);
+                callHandlerCallbackLater(ctx, false);
                 return ctx.handler();
+            }
+            inEventLoop = executor.inEventLoop();
+            if (inEventLoop) {
+                replace0(ctx, newCtx);
+            }
+        }
+        if (inEventLoop) {
+            // Invoke newHandler.handlerAdded() first (i.e. before oldHandler.handlerRemoved() is invoked)
+            // because callHandlerRemoved() will trigger inboundBufferUpdated() or flush() on newHandler and those
+            // event handlers must be called after handlerAdded().
+            callHandlerAdded0(newCtx);
+
+            EventExecutor eventExecutor = ctx.executor();
+            if (eventExecutor.inEventLoop())  {
+                callHandlerRemoved0(ctx);
             } else {
-                future = newCtx.executor().submit(new OneTimeTask() {
+                waitForFuture(eventExecutor.submit(new OneTimeTask() {
                     @Override
                     public void run() {
-                        synchronized (DefaultChannelPipeline.this) {
-                            replace0(ctx, newCtx);
-                        }
+                        callHandlerRemoved0(ctx);
                     }
-                });
+                }));
             }
+        } else {
+            waitForFuture(executor.submit(new OneTimeTask() {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Reduce this anonymous class number of lines from 24 to at most 20, or make it a named class. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1188)